### ISSUE TITLE
Configurable thresholds for error rates

### ISSUE THREAD

**Issue by lucasponce** [kiali/kiali#1424]
As a user I want to be able to configure the error rate thresholds so that they display an appropriate status in health.

I want also to be able to configure different thresholds depending on the service. Ex: service A may still be considered as OK with 5% errors whereas service B is not OK above 1%.

JIRA:
https://issues.jboss.org/browse/KIALI-579

**Comment by jshaughn** [comment]
@lucasponce I'm somewhat surprised no one has upvoted this but perhaps this can be closed given that it is stale and no one has asked for it.  Maybe users are happy enough with the fixed settings.

**Comment by lucasponce** [comment]
@jshaughn +1
Let's re-open if we have more feedback about it.

**Comment by kenske** [comment]
I'd be very interested in having this functionality. It is normal and expected for one of our APIs to return a high percentage of 4xx errors (around 12%). Right now this is always marked as being over the fixed threshold. If we ever do get an abnormal amount, it won't be easily detectable.

**Comment by jotak** [comment]
@kenske I guess that in this situation you would like to make a difference between 4xx and 5xx errors, no? Like having one threshold for 4xx and another for 5xx and they are checked separately?
Also, would it be on a per-service basis?

**Comment by kenske** [comment]
@jotak Correct, having around 12% 4xx errors for our API service is fine, but having any 5xx errors would be a problem. For other services in our cluster, having anything over 1% for 4xx would be a problem.

**Comment by lucasponce** [comment]
@kenske thanks for the feedback :-).
Having a comment from users is enough reason to keep it open.

**Comment by jshaughn** [comment]
I have nothing against a WIP to help flesh out ideas but I think this is a pretty big topic that can probably benefit from more up-front exposure and definition.  Moving from two simple, fixed values to a fully configurable namespace/object/errorCode granularity is pretty big.  We need to make sure that changes here are applied consistently across the affected areas.  For example, the thresholds are not used explicitly for health calculation but also have shared use in the graph when determining edge color.

A couple of notes/questions:
- Will there still be global settings like what we have today, as the defaults?  Meaning, are the detailed configs applied as overrides?
- Will it be possible to inspect/configure the settings from the UI?
- Remember that not everything is HTTP and different protocols have different error codes (or no error codes for TCP).
- The graph actually has a table-driven approach to handling protocols to help it calculate error rates.  Today there is **no support** for code-specific error rate, just a single error rate.  Meaning, today for HTTP 4xx and 5xx is combined.  So, work to do here to support the idea being presented.
  - Also note that the error rate is calculated server-side but the color-selection is done client-side.  So we really need to think about how this would work.

**Comment by lucasponce** [comment]
@jshaughn I will jump on this in the context of the scope for this one.

> Will there still be global settings like what we have today, as the defaults? Meaning, are the detailed configs applied as overrides?

It should be +1.

> Will it be possible to inspect/configure the settings from the UI?

I think this is out of the scope of this preliminar work, at the moment it's being used the configuration as first point. In the future it can be a good point to discuss.

> Remember that not everything is HTTP and different protocols have different error codes (or no error codes for TCP).

+1

> The graph actually has a table-driven approach to handling protocols to help it calculate error rates. Today there is no support for code-specific error rate, just a single error rate. Meaning, today for HTTP 4xx and 5xx is combined. So, work to do here to support the idea being presented.
Also note that the error rate is calculated server-side but the color-selection is done client-side. So we really need to think about how this would work.

In short, this work needs to be consistency, it can't be that the health service shows a different status used from the graph.

The idea is the user can start adding some configuration from the backend, mostly generic for whole scenarios, this should be a base of future refinaments and iterations.

But it's out of the scope of this work to define a fine level granularity for a single object that user can touch/modify from the UI.

**Comment by lucasponce** [comment]
> It's fine to scope out anything complex for inspecting the settings but it is important to be able to explain why health, or edge color, is like it is. Today in the graph legend we do display the fixed thresholds. Maybe initially the user will need to inspect config from the about box.

+1

Yes, good point, this info can be exposed as the config visible from the UI and used to this scenario.

**Comment by jshaughn** [comment]
I had another thought on this that could be a simplifier on the graph-side of things.  As mentioned above, the assigned edge color is based on the edge's error rate as calculated serverside in the graph generation.  The error rate is compared client-side to the thresholds set in config.  This could be quite difficult to do with the fine-grained thresholding proposed here for health.  Instead, perhaps things could be more simple if the edge color was determined not by the error rate set on the edge (which, for example, combines 4xx+5xx for http) but by using the Health object of the target node, perhaps defaulting to the current mechanism only if the health info is not available.

One thing though, it would need to match only the request health, not things like pod health, as the edges reflect traffic only.

I haven't looked much at the code but this approach may actually be doable today, and outside of the actual change to health calculation proposed here.  If it sounds good and I could look into a POC, thus reducing the scope of the work for this proposal.

**Comment by aljesusg** [comment]
I thought this like an addition to the health not a replace.

About how to show this in the UI for the graph we can show a warning icon in the edge with a tooltip with the customizable alert information and not change the color.

**Comment by jshaughn** [comment]
> I thought this like an addition to the health not a replace.

+1, as I understand it the fixed thresholds for health remain but can have overrides that are specific to service and response codes.

> About how to show this in the UI for the graph we can show a warning icon in the edge with a tooltip with the customizable alert information and not change the color.

I think I'd prefer to keep the graph showing health as it is today, and as it has been approved by UX.  Node health is indicated by color on the node.  And the detailed health is available in the side-panel for a selected node (via tooltip, like you suggest).  In general we don't have tooltips on the graph itself.

The poblem is really only the edge color because it is determined independently of the request health calculation.  And so if request health is using different thresholds than those used for edge color it could be confusing for the user.  That is why I was thinking we could move to using the node's request health to determine the edge color. That way no matter what logic we apply in the node health, it would carry over to the incoming edge.

One last note, I think we should probably avoid using 'alert' terminology here and stay with just configurable health thresholds, or something like that.

**Comment by jshaughn** [comment]
Replying to my own comment...   After thinking a little more it's actually not possible to use the same health grade on an edge as is assigned to the destination node.   The destination node's traffic health is based on an aggregation of all the traffic, whereas any individual edge is just a part of that aggregation.  Put simply, if two edges lead into a node, and 1 edge has all failures and the other all success, one should be green and one should be red.  The node health will be determined as an aggregate of those two edges.

So that brings up back to the start.  For an edge to properly use configurable thresholds it needs to easily be able to determine the thresholds.  It's not clear to me where that info should come from.  One idea is that the node's  health info stores with it the thresholds it used when determining the health rating.

**Comment by lucasponce** [comment]
Ok.
I have a first deep review to @aljesusg's work in PR https://github.com/kiali/kiali/pull/2807.

Thanks @aljesusg for your proposal, I think I understand it now and it is interesting as it opens a good discussion.

I will try to ellaborate my thoughts on this. 

First, I would like to warn that this topic is complex as @jshaughn comments and also there are some related epics that are related, so in other words, this comment will be longer than usual :-).

The goal of this original issue is to add the ability to the user to define in a global way how health should be showed in Kiali (in practice, when we mark an element healthy or not).

Of course, this is a topic that can fill a PhD of Computer Science and probably there were different valid opinions, I would much rather be practical and try to limit the scope for Kiali and what we calculate today and how that calculation can be improved via configuration.

From the https://github.com/kiali/kiali/pull/2807 there are some ideas that I really like, but others that I would propose in a different way:

- I would try to change the terminology/naming, I know this is given in the issue, but I would like to call this "Health Config" rather than "Threshold". Why ? There is a clash in terminology with alerting engines and that can bring some confusion about feature itself. Also, in our plans we would like to incorporate other signals to health calculation and alerting https://github.com/kiali/kiali/issues/1416 would be included there.

- I like the idea to have a Health Config defined per namespace, application, workload or service, and also per error code 4XX or 5XX, but I would avoid go beyond that granularity, otherwise we are building some sort of proto-alerting and if the user needs that I think it would be best to define alerts in other system and let Kiali to read the alerting signals.

- This Health Config should be part of the external config that the UI can read, that would help Graph to sync and also apply same Health Config (I will not call it "rules"/"thresholds" to not fall in the overlapping terminology).

- Today, Kiali offers health service per namespace, application, service, and workload. That is good. All error metrics are aggregated, one suggestion I have is to provide a minimal split between codes 4XX, 5XX (all together, no indifivually, only two groups: 4XX (400-499), and 5XX (500-599). So, from the backend then I will get that split, but I won't calculate those complex filtering in the backend.

- As it happens today, the final Health is calculated and printed in the UI, so, if the UI can read the HealthConfig it can apply what health should show. Also, in the graph, it can read the same config to apply it into the same computation.

So, in short, today, we have:

```
GET /namespaces/{namespace}/health
GET /namespaces/{namespace}/apps/{app}/health
GET /namespaces/{namespace}/services/{service}/health
GET /namespaces/{namespace}/workloads/{workload}/health
```

Where errors are aggregated, I think that we can split them with two groups 4XX, 5XX (initially, I want just define a first scenario that can be extended in future iterations).

So, in the Health Configuration I may have simple entries combining namespace/app/service/workload to allow the error tolerance for 4XX or 5XX that will be used by UI in Health.ts and Graph to show the Health Status.

I would not enter in low level filtering of prometheus metrics at backend, for two reasons, I think that is going to add more overload and also it can easily move out of the scope of the Health and enter in what an alerting engine is used for.

Note, having this initial Health Config is something that I expect it can expand when more signals are used. 

In our roadmap we are going easily to add tracing here, some we would like to config something like the "tracing tolerance" before to show that something is unhealthy.

The same will happen with other signals like Alerting.

I see a lot of good comments and proposal, first in @aljesusg PR and also in @jotak and @jshaughn comments. 

I think we can reuse part of this work and move in the direction that I comment but I would like to hear your comments as well.

**Comment by aljesusg** [comment]
> Ok.
> I have a first deep review to @aljesusg's work in PR #2807.
> 
> Thanks @aljesusg for your proposal, I think I understand it now and it is interesting as it opens a good discussion.
> 
> I will try to ellaborate my thoughts on this.
> 
> First, I would like to warn that this topic is complex as @jshaughn comments and also there are some related epics that are involved, so in other words, this comment will be longer than usual :-).
> 
> The goal of this original issue is to add the ability to the user to define in a global way how health should be showed in Kiali (in practice, when we mark an element healthy or not).
> 
> Of course, this is a topic that can fill a PhD of Computer Science and probably there were different valid opinions, I would much rather be practical and try to limit the scope for Kiali and what we calculate today and how that calculation can be improved via configuration.
> 
> From the #2807 there are some ideas that I really like, but others that I would propose in a different way:
> 
> * I would try to change the terminology/naming, I know this is given in the issue, but I would like to call this "Health Config" rather than "Threshold". Why ? There is a clash in terminology with alerting engines and that can bring some confusion about feature itself. Also, in our plans we would like to incorporate other signals to health calculation and alerting #1416 would be included there.
> * I like the idea to have a Health Config defined per namespace, application, workload or service, and also per error code 4XX or 5XX, but I would avoid go beyond that granularity, otherwise we are building some sort of proto-alerting and if the user needs that I think it would be best to define alerts in other system and let Kiali to read the alerting signals.
> * This Health Config should be part of the external config that the UI can read, that would help Graph to sync and also apply same Health Config (I will not call it "rules"/"thresholds" to not fall in the overlapping terminology).
> * Today, Kiali offers health service per namespace, application, service, and workload. That is good. All error metrics are aggregated, one suggestion I have is to provide a minimal split between codes 4XX, 5XX (all together, no indifivually, only two groups: 4XX (400-499), and 5XX (500-599). So, from the backend then I will get that split, but I won't calculate those complex filtering in the backend.
> * As it happens today, the final Health is calculated and printed in the UI, so, if the UI can read the HealthConfig it can apply what health should show. Also, in the graph, it can read the same config to apply it into the same computation.
> 
> So, in short, today, we have:
> 
> ```
> GET /namespaces/{namespace}/health
> GET /namespaces/{namespace}/apps/{app}/health
> GET /namespaces/{namespace}/services/{service}/health
> GET /namespaces/{namespace}/workloads/{workload}/health
> ```
> 
> Where errors are aggregated, I think that we can split them with two groups 4XX, 5XX (initially, I want just define a first scenario that can be extended in future iterations).
> 
> So, in the Health Configuration I may have simple entries combining namespace/app/service/workload to allow the error tolerance for 4XX or 5XX that will be used by UI in Health.ts and Graph to show the Health Status.
> 
> I would not enter in low level filtering of prometheus metrics at backend, for two reasons, I think that is going to add more overload and also it can easily move out of the scope of the Health and enter in what an alerting engine is used for.
> 
> Note, having this initial Health Config is something that I expect it can expand when more signals are used.
> 
> In our roadmap we are going easily to add tracing here, some we would like to config something like the "tracing tolerance" before to show that something is unhealthy.
> 
> The same will happen with other signals like Alerting.
> 
> I see a lot of good comments and proposal, first in @aljesusg PR and also in @jotak and @jshaughn comments.
> 
> I think we can reuse part of this work and move in the direction that I comment but I would like to hear your comments as well.

I undestand your proposal but I see so useful have a label and prometheus filter. About what kind of checks... In first step I thought in the error code 4xx 5xx but I think that it could be useful have rules for other prometheus properties to check requests where destination is X for example.

**Comment by lucasponce** [comment]
> I undestand your proposal but I see so useful have a label and prometheus filter. About what kind of checks... In first step I thought in the error code 4xx 5xx but I think that it could be useful have rules for other prometheus properties to check requests where destination is X for example.

That is not a bad feature, but when the granularity is like you describe, that is an alerting engine.
And that is out of the scope for the moment, there is a better place to define rules including complex expressions (service, destination, error code; combinations of all of this).

On the contrary, what I try to express in my previous comment is that today there is not a "tolarence" level in some properties that I guess there is room to configure them without trying to replicate what a proper alerting engine is doing.

Even, I guess in terms of heatlh config, I re-think about it and I think it should be something more close to namespace granulartity rather than service/workload/app.

Also, when we introduce more signals: next is tracing, in the roadmap will be signals processed from logs, or other istio events, then we may "configure" when something is health (when tracing throws errors some users may want to put a "unhealthy", but others a "degraded"; when some workload has some alerting defined, probably user want to see a "unhealthy").

To me these two concepts are somehow related but are different. If we think in terms of "Health Config" I think details are clear compared with such level of granularity that you propose.

**Comment by aljesusg** [comment]
ok so please we need to discuss the target and the definition of the task. I don't want to invest time if we are going to change it after. Can we discuss this in the issue or you prefer a bj?

**Comment by lucasponce** [comment]
Let's review this one tomorrow after the scrum meeting.
I think that @jotak was taking some working notes in a document that we can use.
In any way, we will post here the summary and details.

**Comment by lucasponce** [comment]
@aljesusg @jotak @xeviknal and me we have discussed this issue and put in context with following efforts related to health. Most of the comments in https://github.com/kiali/kiali/issues/1424#issuecomment-642573600 are still valid.

A summary of main ideas that would drive this issue:

- Kiali calculates health correlating signals: pods status, Istio error rates. In next iterations more signals will be included: tracing errors, log erros and alerting.
- For alerting, there are some open issues in Kiali https://github.com/kiali/kiali/issues/1270 https://github.com/kiali/kiali/issues/1416 but the plans are to incorporate them in mid/long term as other signal that can be used to correlate and health calculation.
- Kiali would provide a way to add a Health Config capability, in a way that user can define some level of configuration about when an element is marked as "healthy", "degraded" or "unhealthy". This configuration may take into consideration all signals used (pods, error request ratio, future tracing/logging/alerting integration).
- Health Config will be available in the UI, in a way that can be queried from "About" feature from the user and used to apply health in graph elements and kiali list/details.
- We have discussed some implementation details (to provide all error requests grouped by code or individually by code) or which level of granularity we want to apply health config (namespace, app, workload or service). But we have an agreement to not apply label filtering on prometheus time series. On that level, Kiali will wait until it can be properly integrated with an existing alerting engine.

@aljesusg will continue working on this effort on a new PR where we can discuss/review more details related to implementation.

**Comment by jotak** [comment]
Ask for 4xx / 5xx distinction for error detection & edge colors on Graph: https://github.com/kiali/kiali/issues/2991

**Comment by aljesusg** [comment]
> User ask for 4xx / 5xx distinction for error detection & edge colors on Graph: #2991.
> 
> This makes me think: this would be even better if we could tell: "Consider all my 4xx as errors when they come from a known workload; and not as errors when they come from gateway / unknown workload". What do you think (@aljesusg @lucasponce @jshaughn ) ? It doesn't change anything for the current work @aljesusg is doing, but could be something to add to the epic?
> 
> The rationale is that a service opened to mesh-external requests shouldn't be considered unhealthy when client requests are badly formed.
> 
> EDIT (thinking again about it): 4xx should be treated by default as errors for _outbound_, and 5xx for _inbound_ ?

The PR about configuration is ready to review and merge, next steps is improve it and be sure the graph edge work with this status

**Comment by jshaughn** [comment]
@jotak While I agree that Client errors generated from internal requests could likely be more significant than those from external requests, I think it's probably too much for us to get into that level of granularity.   At least for now and not without community ask.  I expect the fact that the health of the known workload being degraded could be enough of a signal.

@aljesusg We're going to really have to think about how to extend this to graph edges.  It's going to have an impact on E2E graph code.

**Comment by aljesusg** [comment]
> @jotak While I agree that Client errors generated from internal requests could likely be more significant than those from external requests, I think it's probably too much for us to get into that level of granularity. At least for now and not without community ask. I expect the fact that the health of the known workload being degraded could be enough of a signal.
> 
> @aljesusg We're going to really have to think about how to extend this to graph edges. It's going to have an impact on E2E graph code.

About that...I am checking that in the Graph you use the health defined in types/health so this should work for graph too, I need to check it. (This is related with nodes) Other thing is about how you show this health in edges (Researching)

```tsx
 private updateNodeHealth(
    ele: Cy.NodeSingular,
    promise: Promise<H.NamespaceAppHealth | H.NamespaceServiceHealth | H.NamespaceWorkloadHealth>,
    key: string
  ) {
    ele.data(
      'healthPromise',
      promise.then(nsHealth => nsHealth[key])
    );
    promise
      .then(nsHealth => {
        // Discard if the element is no longer valid
        if (this.isElementValid(ele)) {
          const health = nsHealth[key];
          if (health) {
            const status = health.getGlobalStatus();
            ele.removeClass(H.DEGRADED.name + ' ' + H.FAILURE.name);
            if (status === H.DEGRADED || status === H.FAILURE) {
              ele.addClass(status.name);
            }
          } else {
            ele.removeClass(`${H.DEGRADED.name}  ${H.FAILURE.name} ${H.HEALTHY.name}`);
            console.debug(`No health found for [${ele.data(CyNode.nodeType)}] [${key}]`);
          }
        }
      })
      .catch(err => {
        // Discard if the element is no longer valid
        if (this.isElementValid(ele)) {
          ele.removeClass(`${H.DEGRADED.name}  ${H.FAILURE.name} ${H.HEALTHY.name}`);
        }
        console.error(`Could not fetch health for [${ele.data(CyNode.nodeType)}] [${key}]: ${API.getErrorString(err)}`);
      });
  }
```

**Comment by jshaughn** [comment]
@aljesusg, @jotak  The way the edge health works today is very simple.  It tests the percentErr against the fixed thresholds currently defined for DEGRADED and FAILURE.  As it is today the percentErr for HTTP includes 4xx+5xx.  But we do have available the raw rates for total/3xx/4xx/5xx.    So it is possible to calculate an edge health in a more granular way if we have the edge-specific thresholds.  If the Health objects on the edge's source and destination nodes contained the thresholds, or provided a function to call, we could feed in the granular rates to determine the edge health.
